### ISSUE TITLE
Dummy imports needed to resolve names from imported modules

### ISSUE THREAD

**Issue by ranjitjhala** [ucsd-progsys/liquidhaskell#1688]
@adinapoli and @kosmikus here is a minimized example of the "dummy imports" problem.

- `module Lib` defines a datatype `T` with constructor `C` and either _reflects_ a function foo (or makes another datatype `T2` with ctor `C2`) whose refinement uses C;

- `module Client` imports `Lib` but LH crashes when resolving the reflected definition of `foo` or `C2` because the name `C` is not in scope :-(

We have been working around this by forcibly-importing C by adding a line

dummy = (C, ...)

but this is painful at the least... Any ideas?

## Lib.hs

```haskell
{-# LANGUAGE GADTs #-}

{-@ LIQUID "--reflection"  @-}

module Lib where

{-@ measure propOf :: a -> b @-}
{-@ type ProofOf E = { proofObj:_ | propOf proofObj = E } @-}

data Expr = Bc Bool                   -- True, False
          | Ic Int                    -- 0, 1, 2, ...
          | Lambda Int Expr         -- \x.e          abstractions
          | App Expr Expr             -- e e'          applications

data Basic = TBool         -- Bool
           | TInt          -- Int

data StepProp where
    Step :: Expr -> Expr -> StepProp

data StepProof where
    EFake :: Vname -> Expr -> Expr -> StepProof

{-@ data StepProof where
    EFake :: x:Vname -> e:Expr -> v:_ -> ProofOf( Step (App (Lambda x e) v) e)
  @-}
```

## Client.hs

```haskell
{-# LANGUAGE GADTs #-}

{-@ LIQUID "--reflection"  @-}

module Client where

import Lib

-- PROBLEM: If we comment `dummy` then `Step` is not in scope and the resolution of `EFake` fails.
dummy = (\x y -> Step x y)

data HasFType where
    FTBC   :: Bool -> HasFType

{-@ ftypSize :: HasFType -> { n:Int | n >= 0 } @-}
ftypSize :: HasFType -> Int
ftypSize (FTBC {}) = 1
```

This is unfortunate. 

If we are loading `Lib` then why can't we also load the
constructor name `Step` in automatically? Presumably this constructor/name
also lives in the serialized spec? (Problem is the GHC doesn't know this name
when it is compiling `Client`?)

**Comment by ranjitjhala** [comment]
derived from but not really related to #1687

**Comment by adinapoli** [comment]
Yes, this is indeed unfortunate. I have reproduced this example in a small project to have a better sense for it (I had to type alias `type Vname = Int` to make the code typecheck, tough) and I think I see what's going on.

It seems to me (based on my current knowledge of things) that LH is trying to "resolve" the reflected `StepProof` inside the `Client` module, for some reason (because compiling `Lib.hs` works just fine for me using the plugin) and obviously choking because nothing in `Client` explicitly mentions `Step`. I have two questions/general observations:

1. Is that really the case (that we resolve the reflected `StepProof` inside `Client` and not when we compile `Lib`)? Can we do anything about that? 

2. Generally speaking, I suspect that we are already resolving all the `GHC` names exported by `Lib` (which should include `Step`), but it almost seems like we are not passing them around in the place where we need them for name resolution. Perhaps we can do something about it.

**Comment by adinapoli** [comment]
Looking at the error, I find it quite interesting:

```
[src/Lib.hs:25:5-47: Error: Illegal type specification for `Lib.EFake`
     Lib.EFake :: x:GHC.Types.Int -> e:Lib.Expr -> v:Lib.Expr -> {VV : Lib.StepProof | EFake VV == v
                                                                                       && EFake VV == e
                                                                                       && EFake VV == x
                                                                                       && propOf VV == Step (App (Lambda x e) v) e}
     Sort Error in Refinement: {VV : Lib.StepProof | (Lib.EFake##v VV == v##Lib.EFake
                                                      && Lib.EFake##e VV == e##Lib.EFake
                                                      && Lib.EFake##x VV == x##Lib.EFake
                                                      && (is$Lib.EFake VV <=> true)
                                                      && propOf VV == Step (App (Lambda x##Lib.EFake e##Lib.EFake) v##Lib.EFake) e##Lib.EFake)}
     Unbound symbol Step --- perhaps you meant: len ?]
```

It seems like _some_ names are being resolved qualified (like `EFake`) but things like `Step` or `App` are not. Why is that? It seems like LH "thinks" these are local names. **I am starting to suspect a bug in the name resolution at this point**, which is causing `Step` to be resolved as a "local" variable. This happened in the past, and it seems likely it might be happening here as well, albeit for different causes, maybe.

**Comment by kosmikus** [comment]
In continuation of our discussion on Tuesday, I just tried this with manually expanding the LH type synonym, and indeed, then it works. So it seems like our speculation that something is going wrong in expanding the type synonym wrt name resolution was correct.

**Comment by adinapoli** [comment]
Ah, that's great, thanks Andres. I suspect a _potential_ solution might be to either expand the type synonym sooner _or_ to fully qualify the names in the type synonym so that it doesn't matter when we expand it. I have yet to look at the relevant code, so I have no sense which of the two solution is easier, and whether or not there are other solutions I am not thinking about right now 😉

**Comment by adinapoli** [comment]
Ok, so I have discovered something quite interesting in my investigation so far. It all starts from the [qualifyExpand](https://github.com/ucsd-progsys/liquidhaskell/blob/develop/src/Language/Haskell/Liquid/Bare/Expand.hs#L265) function:

```haskell
qualifyExpand :: (Expand a, Bare.Qualify a) => Bare.Env -> ModName -> BareRTEnv -> F.SourcePos -> [F.Symbol] -> a -> a 
qualifyExpand env name rtEnv l bs
  = expand rtEnv l  . Bare.qualify env name l bs
```

The aim of this function is to essentially resolve names and expand things like type synonyms before creating a `LiftedSpec`.
I have traced the input and output of `expand` and `qualify`, and I have noticed something interesting. Let's observe what happens to `StepProof` if we qualify and later expand. `StepProof` it's a `DataDecl`, which means we expand all their type arguments and return type. Let's recall its refinement type:

```
{-@ data StepProof where
    EFake :: x:Vname -> e:Expr -> v:_ -> ProofOf( Step (App (Lambda x e) v) e)
  @-}
``` 

When we qualify, LH turns that data declaration into something like this (a **note on the debug output**: the arrow in the middle separates the input to the function from the output):
 
```
Trace: [qualify DataDecl data StepProof  [] =
    | EFake :: forall . x : Vname ->e : Expr ->v : {VV : _ | $HOLE} -> (ProofOf (Step (App (Lambda x e) v) e)) ==> ] : 
data StepProof  [] =
    | EFake :: forall . x : Vname ->e : Expr ->v : {VV : _ | $HOLE} -> (ProofOf (Step (App (Lambda x e) v) e))
```

I find this surprising for two reasons:

1. LH is not fully-qualifying `StepProof` or `EFake` at this point in time;
2.  _Other_ data declarations defined in the same module are essentially **already qualified** by the time we call the function. Example:

```
Trace: [qualify DataDecl data T1688Lib.Basic  [] =
    | T1688Lib.TBool :: forall . -> T1688Lib.Basic
    | T1688Lib.TInt :: forall . -> T1688Lib.Basic ==> ] : data T1688Lib.Basic  [] =
    | T1688Lib.TBool :: forall . -> T1688Lib.Basic
    | T1688Lib.TInt :: forall . -> T1688Lib.Basic
```

I wonder why is that -- does calling `Bare.qualify` only qualify refined data declarations?

Regardless of the surprising behaviour above, it's now fairly easy to see how this won't work: we are now expanding something which was really not qualified in the first place, so the expansion becomes:

```
Trace: [inputType => (ProofOf (Step (App (Lambda x e) v) e))] : {proofObj : _ | propOf proofObj == Step (App (Lambda x e) v) e}
```

So LH correctly expanded `ProofOf`, but at this point it's game over as `Step`, `App` and `Lambda` are not qualified.

**Interestingly enough**, if we **flip** the order in which we do operations (i.e. we expand _first_, we qualify _later_), things works correctly. During qualification, we get:

```
Trace: [qualify DataDecl data StepProof  [] =
    | EFake :: forall . x : Vname ->e : Expr ->v : {VV : _ | $HOLE} -> {proofObj : _ | propOf proofObj == Step (App (Lambda x e) v) e} ==> ] : data StepProof  [] =
    | EFake :: forall . x : Vname ->e : Expr ->v : {VV : _ | $HOLE} -> {proofObj : _ | propOf proofObj == T1688Lib.Step (T1688Lib.App (T1688Lib.Lambda x e) v) e}
```

What is interesting once again, though, is that although `Step`, `App` and `Lambda` are now correctly qualified `StepProof` and `EFake` are not. I have no clue if this might lead to problems down the line, but it makes the test pass.

Surprisingly, I would have thought that flipping the order of these two operations was going to lead to a plethora of failing tests, but it seems like the majority still works. The only failures I've got were related to _inline_ definitions (i.e. function marked using the LH's `inline` keyword) but that seems like a case we could handle separately if we want to go down this road.

To recap, this is the current situation:

* For this particular test (I have no data points to generalise) it seems like LH is not doing much when it comes to qualify the `StepProof` data declaration (I will have to found out why, though);
* Inverting the order of expansion & qualification makes the test pass, but I am not fully convinced that's really "the fix".

I will keep investigating and report back.
